lib/ansible/galaxy/dependency_resolution/dataclasses.py

```diff
@@ -415,13 +419,13 @@ def is_online_index_pointer(self):
 
 class Requirement(
         _ComputedReqKindsMixin,
-        namedtuple('Requirement', ('fqcn', 'ver', 'src', 'type')),
+        namedtuple('Requirement', ('fqcn', 'ver', 'src', 'type', 'signature_sources')),
```

Comment:
I was thinking... Is it significant to preserve the sources? Those alternative constructors above accept an artifact manager as an argument, why not use it to get the actual signatures from the sources and construct the instance with the same attribute `signatures` as `Candidate` has? Is this differing `signature_sources` attribute really necessary on this level?

Reply:
It seems signicant and is separate from Candidate.signatures; it's additive. The requirements.yml parsing was moved into the dataclasses, and that's where this information comes from. We also need this information in providers, because otherwise we'd need to iterate over the resolved Candidates again and figure out which Requirements correlate to which Candidates.

Reply:
> Could you expand on why they need to be correlated by a full list scan?

We'd need to scan the Requirements fully again because multiple requirements can correspond to the same Candidate.
```
collections:
  - name: ns.coll
    version: >1.0.0
    signatures: [ ... ]
  - name: ns.coll
    version: <=2.0.0
    signatures: [ ... ]
```
If a matching candidate is found, both of those signature lists apply.

We know the corresponding Requirements in the dep resolver already, because we're using them to find matching Candidates.

An alternative could be saving a new requirements attr on the Candidate and using it after the dep resolver to recreate the Candidate with the Requirement signatures. It just seems unnecessary to create and then recreate the Candidate when it could be done the first time.

> Or why can't the signatures be extracted during the initialization time so that a Requirement instance would contain all of the acceptable signatures when it's constructed?

By "extract", I assume you mean from the Galaxy server. We don't know what server the collection will come from until we resolve the Requirements using the dep resolver. Trying to determine where the collection will come from so we can create a Requirement containing all signatures ahead of time seems inefficient and I'd worry about the code paths diverging.
And we'd still need to use the requirement signatures in providers if we don't want to re-determine which Requirement(s) match which Candidate. I don't see a good way to create the Requirement with all acceptable signatures.